functions/vshn-postgres-func/restart.go

```diff
@@ -0,0 +1,143 @@
+
+// TransformRestart triggers a restart of the postgreqsql instance if there is a pending restart and the composite is configured to restart on update.
+func TransformRestart(ctx context.Context, iof *runtime.Runtime) runtime.Result {
+	return transformRestart(ctx, iof, time.Now)
+}
+
+func transformRestart(ctx context.Context, iof *runtime.Runtime, now func() time.Time) runtime.Result {
+	comp := vshnv1.VSHNPostgreSQL{}
+	err := iof.Desired.GetComposite(ctx, &comp)
+	if err != nil {
+		return runtime.NewFatal(ctx, err.Error())
+	}
+	err = keepRecentRestartOps(ctx, iof, comp.GetName(), now)
+	if err != nil {
+		return runtime.NewFatal(ctx, err.Error())
+	}
+
+	if comp.Spec.Parameters.UpdateStrategy.Type != vshnv1.VSHNPostgreSQLUpdateStrategyTypeOnChange {
+		return runtime.NewNormal()
+	}
+
+	restartTime, err := getPendingRestart(ctx, iof)
+	if err != nil {
+		return runtime.NewWarning(ctx, err.Error())
+	}
+
+	if restartTime.IsZero() {
+		return runtime.NewNormal()
+	}
+
+	err = scheduleRestart(ctx, iof, comp.GetName(), restartTime)
+	if err != nil {
+		return runtime.NewFatal(ctx, err.Error())
+	}
+
+	return runtime.NewNormal()
+}
+
+func getPendingRestart(ctx context.Context, iof *runtime.Runtime) (time.Time, error) {
+	cluster := sgv1.SGCluster{}
+
+	err := iof.Observed.GetFromObject(ctx, &cluster, "cluster")
+	if err != nil {
+		return time.Time{}, err
+	}
+
+	log := controllerruntime.LoggerFrom(ctx).WithValues("cluster", cluster.Name)
+	if cluster.Status.Conditions == nil {
+		return time.Time{}, nil
+	}
+
+	for _, cond := range *cluster.Status.Conditions {
+		if cond.Type == nil || *cond.Type != sgv1.SGClusterConditionTypePendingRestart || cond.Status == nil || cond.LastTransitionTime == nil {
+			continue
+		}
+		status, err := strconv.ParseBool(*cond.Status)
+		if err != nil || !status {
+			continue
+		}
+
+		log.WithValues("at", cond.LastTransitionTime).Info("PendingRestart")
+
+		restartTime, err := time.Parse(time.RFC3339, *cond.LastTransitionTime)
+		if err != nil {
+			continue
+		}
+		return restartTime, nil
+	}
+
+	return time.Time{}, nil
+}
+
+func scheduleRestart(ctx context.Context, iof *runtime.Runtime, compName string, restartTime time.Time) error {
+	name := fmt.Sprintf("pg-restart-%d", restartTime.Unix())
+	ops := sgv1.SGDbOps{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: fmt.Sprintf("vshn-postgresql-%s", compName),
+		},
+		Spec: sgv1.SGDbOpsSpec{
+			Op:        sgv1.SGDbOpsOpRestart,
+			SgCluster: compName,
+			Restart: &sgv1.SGDbOpsSpecRestart{
+				Method:             pointer.String(sgv1.SGDbOpsRestartMethodInPlace),
+				OnlyPendingRestart: pointer.Bool(true),
+			},
+			RunAt: pointer.String(restartTime.Format(time.RFC3339)),
+		},
+	}
+	return iof.Desired.PutIntoObject(ctx, &ops, fmt.Sprintf("%s-%s", compName, name))
+}
+
+func keepRecentRestartOps(ctx context.Context, iof *runtime.Runtime, compName string, now func() time.Time) error {
+
+	for _, r := range iof.Observed.List(ctx) {
```

Comment:
Does the observed list actually contain the SgDbOps objects? See my observation here: https://github.com/crossplane/crossplane/issues/3918#issuecomment-1493775378

Reply:
I will triple check, but as far as I can tell, yes it works.

Crossplane "remembers" the resources for one iteration. So whatever is in the "desired" will show up in the "observed" in the next iteration. That's why I copy all SgDbOps objects that I still need from  "observed" to "desired"

Reply:
yeah I also had this doubt.

Reply:
> I will triple check, but as far as I can tell, yes it works.
> 
> Crossplane "remembers" the resources for one iteration. So whatever is in the "desired" will show up in the "observed" in the next iteration. That's why I copy all SgDbOps objects that I still need from "observed" to "desired"

Good, then it's probably only the `desired` that doesn't behave as described in the docs. Just wanted to be sure we don't accidentally run into something weird here...